File: java/src/main/java/com/example/cote/GJM/알러지가_심한_친구들.java
package com.example.cote.GJM;

import java.nio.file.*;
import java.io.*;
import java.text.*;
import java.util.*;
import java.util.stream.*;

public class 알러지가_심한_친구들 {
	
	private final static int N = 20;
	private final static int M = 20;
	private final static int INF = 987654321;
	private static int n;
	private static int m;
	private static int best;
	private static List<Integer>[] eater;
	private static List<Integer>[] canEat;


	public static void initialize(BufferedReader fd) throws IOException{
		
		Map<String,Integer> nameToIdx = new HashMap<> ();

		eater = new LinkedList[m];
		for(int i =0; i<m; i++){
			eater[i] = new LinkedList<>();
		}

		canEat = new LinkedList[n];
		for(int i=0; i<n; i++){
			canEat[i] = new LinkedList<>();
		}
		

		String[] names = fd.readLine().split(" ");
		
		IntStream.range(0,names.length)
			.forEach(i->{
				nameToIdx.put(names[i],i);
			});


		for(int i=0; i<m; i++){
			String[] tokens = fd.readLine().split(" ");

			for(int t=1; t<tokens.length; t++){
				int p = nameToIdx.get(tokens[t]);
				eater[i].add(p);
				canEat[p].add(i);
			}			
		}		

	}

	private static void searchSlow(int cur, List<Integer> edible, int chosen){
		if(best <= chosen) return;
		if( cur == m ){

			if(edible.stream().noneMatch(i->i==0))
				best = Math.min(best, chosen);
			return;
		}

		searchSlow(cur+1, edible, chosen);

		eater[cur].stream().forEach(
			i-> {edible.set(i, edible.get(i)+1);}
		);
		searchSlow(cur+1, edible, chosen+1);	
		eater[cur].stream().forEach(
			i-> {edible.set(i, edible.get(i)-1);}
		);

	}

	private static int solveSlow(){
		best = INF;
		List<Integer> edible = new LinkedList<>();
		for(int i=0; i<n; i++) edible.add(0);
		searchSlow(0, edible, 0);

		return best;
	}

	private static void search(List<Integer> edible, int chosen){
		
		if( best <= chosen ) return;

		OptionalInt first = IntStream.range(0,n)
							.filter(i->edible.get(i)==0)
							.findFirst();
							
		if (!first.isPresent()){
			best = Math.min(best,chosen);
			return; }

		int p = first.getAsInt();

		for(int i =0; i<canEat[p].size(); i++){
			int cur = canEat[p].get(i);

			eater[cur].stream().forEach( 
				j -> {edible.set(j, edible.get(j)+1);}
			);

			search(edible, chosen+1);

			eater[cur].stream().forEach( 
				j -> {edible.set(j, edible.get(j)-1);}
			);

		}
		



	}

	private static int solve(){
		best = INF;
		List<Integer> edible = new LinkedList<>();
		for(int i=0; i<n; i++) edible.add(0);

		search(edible, 0);
		return best;
	}
	
	public static void main(String... args) throws IOException, ParseException{

		Path p = Paths.get(System.getProperty("user.dir")+"/data/알러지가_심한_친구들.txt");
		MessageFormat mf = new MessageFormat("{0} {1}");
		BufferedReader fd = Files.newBufferedReader(p);

		int testCases = Integer.parseInt(fd.readLine());

		while(testCases > 0){

			int[] nums = Arrays.stream(mf.parse(fd.readLine()))
							.map(String::valueOf)
							.mapToInt(Integer::parseInt)
							.toArray();

			n = nums[0];
			m = nums[1];

			initialize(fd);

			System.out.println(solveSlow()+","+ solve());

			testCases--;

		}
		
	}
}

File: kotlin/src/main/kotlin/com/example/cote/GJM/알러지가_심한_친구들.kt
package com.example.cote.GJM

import java.io.*
import java.nio.file.*

object 알러지가_심한_친구들 {

    val N = 20
    val M = 20
    val INF = 987654321
    var best = INF

    fun search(
        eater: List<List<Int>>,
        canEat: List<List<Int>>,
        edible: MutableList<Int>,
        n: Int,
        m: Int,
        chosen: Int
    ) {

        if (chosen >= best) {
            return
        }

        val allSelected = edible.all { it > 0 }

        if (allSelected) {
            best = Math.min(best, chosen)
            return
        }

        val cur = edible.withIndex().filter { it.value == 0 }.first().index

        for (food in canEat[cur]) {

            for (person in eater[food]) {
                edible[person] += 1
            }

            search(eater, canEat, edible, n, m, chosen + 1)

            for (person in eater[food]) {
                edible[person] -= 1
            }
        }
    }

    fun solve(
        eater: List<List<Int>>,
        canEat: List<List<Int>>,
        n: Int,
        m: Int
    ): Int {
        best = INF

        search(eater, canEat, MutableList(n) { 0 }, n, m, 0)

        return best
    }

    fun main() {
        val p =
            Paths.get(System.getProperty("user.dir") + "/data/알러지가_심한_친구들.txt")
        val reader = Files.newBufferedReader(p)

        repeat(reader.readLine().toInt()) {
            val l = reader.readLine().split(" ").map(String::toInt)
            val n = l[0]
            val m = l[1]

            val person = listOf(*reader.readLine().split(" ").toTypedArray())
            val eater = List(m) { mutableListOf<Int>() }
            val canEat = List(n) { mutableListOf<Int>() }
            repeat(m) {
                val token =
                    reader
                        .readLine()
                        .trim()
                        .split(" ")
                        .drop(1)
                        .map(person::indexOf)

                eater[it].addAll(token)

                for (v in token) {
                    canEat[v].add(it)
                }
            }

            println(solve(eater, canEat, n, m))
        }

        reader.close()
    }
}

fun main() {
    알러지가_심한_친구들.main()
}
